### PR TITLE
 Add logging, non blocking receive 

### PR DESCRIPTION
**Pull request by nicoschoenmaker**
- Additional log statements make it easier to debug potential protocol errors
- The receive loop is now non-blocking, allowing the capability of a timeout

### CODE REVIEW

File: src/Bundler/Runner/UnixSocket.php [hunk 1]
```diff
@@ -18,25 +20,32 @@
...
         }
     }
 
     public function send(string $to_send, int $bytes): void
     {
-        if (! @socket_send($this->socket, $to_send, $bytes, 0)) {
+        $this->logger->debug('[UnixSocket] Sending ' . $bytes);
+        if (!@socket_send($this->socket, $to_send, $bytes, 0)) {
+            $this->logger->debug('[UnixSocket] Sending exception');
             throw new SocketException('Problem sending to the socket');
```
Comment (iltar) [inline src/Bundler/Runner/UnixSocket.php]: What if we do: `throw SocketException::send($this->socket)`

File: src/Bundler/Runner/UnixSocket.php [hunk 2]
```diff
@@ -46,10 +55,24 @@ public function receive(int $bytes): string
...
+
+        while ($bytes > $length) {
+            $this->logger->debug('[UnixSocket] Reading ' . ($bytes - $length) . ' bytes');
+            $res = @socket_read($this->socket, $bytes - $length);
+            if ($res === false) {
+                $this->logger->debug(
+                    '[UnixSocket] Reading exception '
+                    . socket_last_error() . ' (' . socket_strerror(socket_last_error()) . ')'
+                );
+                throw new SocketException('Problem receiving from the socket');
```
Comment (iltar) [inline src/Bundler/Runner/UnixSocket.php]: What if we do `throw SocketException::read($this->socket)`

File: src/Bundler/Runner/UnixSocket.php [hunk 3]
```diff
@@ -18,25 +20,32 @@
...
         $this->socket = socket_create(AF_UNIX, SOCK_STREAM, 0);
+        socket_set_option($this->socket, SOL_SOCKET, SO_RCVTIMEO, ['sec' => 3, 'usec' => 0]);
     }
 
     public function connect(string $address): void
     {
-        if (! @socket_connect($this->socket, $address)) {
+        if (!@socket_connect($this->socket, $address)) {
+            $this->logger->debug('[UnixSocket] Connection exception');
             throw new SocketException('Problem connecting to the socket');
```
Comment (iltar) [inline src/Bundler/Runner/UnixSocket.php]: What if we do `throw SocketException::connect($this->socket)` with `socket_strerror(socket_last_error())`?

We should probably catch the exception and _then_ log where we catch it. Right now we're logging _and_ throwing, which seems kinda cumbersome as it mainly logs and then throws. My suggestion is basically to push the logger dependency out of the unix socket and into the layers calling/catching this.

Same goes for all the other exceptions.

File: src/Bundler/Runner/UnixSocketRunner.php [hunk 4]
```diff
@@ -31,24 +32,28 @@ class UnixSocketRunner implements RunnerInterface
...
         UnixSocketFactory $factory,
-        int $start_timeout = 500000,
+        LoggerInterface $logger = null,
+        int $start_timeout = 5000000,
         int $small_timeout = 100000
     ) {
         $this->config          = $config;
         $this->socket_location = $config->getCacheDir() . '/asset-lib.socket';
         $this->factory         = $factory;
+        $this->logger          = $logger ?? new NullLogger();
```
Comment (yannickl88) [inline src/Bundler/Runner/UnixSocketRunner.php]: doesn't the logger also exists in the config? `$config->getLogger()`

Comment (nicoschoenmaker) [reply src/Bundler/Runner/UnixSocketRunner.php]: Good point :+1:

**Pull request merged**